Elasticity Patch PA

justinlaughlin opened the pull request:
This PR continues the work in #3088 by extending patch-based integration to elasticity.

Primary changes:
 - Patch PA for linear elasticity
   - New miniapp `nurbs_patch_ex2` 
   - PA Jacobi
 - Some generalizations for IGA/patch-based routines
   - `PatchBasisInfo` - holds `B`, `G`, and other related data used for patch integration
   - `PatchG3D` and `PatchGT3D` - `G` and `G^T` PA operators using sum-factorization (`dim` is fixed, `vdim` is templated)
   - `SplineIntegrationRule` - switch for different "spline level" integration behavior. This could probably be combined with the optimized reduced integration rules, but right now it just defines "simple" rules.
     - This fixes a bug in `nurbs_patch_ex1`; if `incdeg` is used but `-iro` is not specified to the correct order, the wrong number of quadrature points will be used (`fec->GetOrder()` does not seem to update, and is not enough info anyways if orders differ per dimension). Now the number of quadrature points are set by using knotvector order.
   - `ppa_data` - `std::vector<Vector>` for holding patch PA data (fixes a bug where the same `pa_data` was being applied to each patch)
 - Some minor speedups. Running `./nurbs_patch_ex1 -ref 4 -incdeg 2 -iro 6 -patcha -pa -fint -no-compare` on my machine:
   - ~66s on `master`
   - ~10s on this branch

When the PR for `mfem-lor-nurbs` comes along, I'll update `nurbs_patch_ex2` (and ex1, if we want?) to have the option of LOR AMG preconditioning.

<!--GHEX{"author":"justinlaughlin","assignment":"2025-09-23T21:33:34","editor":"tzanio","id":4727,"reviewers":["dylan-copeland","YohannDudouit"],"merge":"2025-10-14T21:33:34","approval":"2025-10-07T21:33:34"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4727](https://github.com/mfem/mfem/pull/4727) | @justinlaughlin | @tzanio | @dylan-copeland + @YohannDudouit | 9/23/25 | ⌛due 10/7/25 | ⌛due 10/14/25 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

justinlaughlin committed:
change Read/ReadWrite to HostRead/HostReadWrite

justinlaughlin committed:
Naming

justinlaughlin committed:
Add skeleton of methods to implement

justinlaughlin committed:
Skeleton of ex2 with patch partial assembly

justinlaughlin committed:
add not implemented error

justinlaughlin committed:
rewrote example in parallel to use hypreboomeramg

justinlaughlin committed:
cleanup and attempting PA

justinlaughlin committed:
add AssembleNURBSPA

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Elasticity patch PA wip

justinlaughlin committed:
Elasticity patch partial assembly

justinlaughlin committed:
Elasticity AddMultPatchPA wip

justinlaughlin committed:
Update nurbs_patch_ex2.cpp

justinlaughlin committed:
Elasticity AddMultPatchPA wip

justinlaughlin committed:
Use DeviceTensor (via Reshape) instead of vector<vector<Array3D>> for holding intermediate values

justinlaughlin committed:
Interpolate grad_uhat at quadrature points

justinlaughlin committed:
Interpolate lambda/mu at quadrature pts for patches

justinlaughlin committed:
Compute stress

justinlaughlin committed:
Some comments on sum factorization and beginning contraction with test function

justinlaughlin committed:
Fixed error in stress calculation (forgot w*detJ); more work on contraction with grad_v

justinlaughlin committed:
Optimize stress calculation

justinlaughlin committed:
Finish contraction for AddMult

justinlaughlin committed:
Fix typo

justinlaughlin committed:
Mistake in div term

justinlaughlin committed:
Update AddMult(); added J^{-1} and fixed some errors

justinlaughlin committed:
Elasticity patch PA...work in progress

justinlaughlin committed:
Fixed bug - pa_data was only using data from last patch

justinlaughlin committed:
Interpolate over all dofs/quads

justinlaughlin committed:
Test cases for nurbs

justinlaughlin committed:
naming

justinlaughlin committed:
mark boundaries

justinlaughlin committed:
typo

justinlaughlin committed:
simplify test mesh

justinlaughlin committed:
AddMult is working now

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Cleanup; move PatchBasisData into its own struct

justinlaughlin committed:
Cleanup

justinlaughlin committed:
optimize AddMult so shared memory is used; More cleanup

justinlaughlin committed:
Cleanup and add timing to nurbs_patch_ex2

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Fix copyright

justinlaughlin committed:
cleanup example

justinlaughlin committed:
run style

justinlaughlin committed:
Separate gradient interpolation into its own subroutine

justinlaughlin committed:
Separate kernel application into its own subroutine

justinlaughlin committed:
factor W*detJ into coefficients

justinlaughlin committed:
Separate ApplyTestFunction into its own subroutine

justinlaughlin committed:
Begin adding diagonal for elasticity

justinlaughlin committed:
update miniapps

justinlaughlin committed:
use tensors to make kernel more concise

justinlaughlin committed:
Change grad_uhat to DeviceTensor + minor cleanup

justinlaughlin committed:
Implement AssembleDiagonal for Patch PA Elasticity

justinlaughlin committed:
Move check for patchwise integration into its own method

justinlaughlin committed:
testing lor preconditioning

justinlaughlin committed:
add SOL operator for benchmarking

justinlaughlin committed:
Working towards some benchmarks with LOR vs jacobi PC

justinlaughlin committed:
fix typo

justinlaughlin committed:
Updating example / mesh_gen

justinlaughlin committed:
Updating example / mesh_gen

justinlaughlin committed:
Fix bug related to knot interpolation

justinlaughlin committed:
LOR pc in example

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Add helper function KnotVector.GetUniqueKnots()

justinlaughlin committed:
Implement Reduced Gaussian Quadrature Rule

justinlaughlin committed:
Update example

justinlaughlin committed:
Update example

justinlaughlin committed:
Cleanup

justinlaughlin committed:
hypre amg for LOR PC - wip

justinlaughlin committed:
Working AMG for LOR PC; cleanup example

justinlaughlin committed:
Renaming

justinlaughlin committed:
Apply style

justinlaughlin committed:
minor

justinlaughlin committed:
typo

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Cleanup - spline integration rules

justinlaughlin committed:
Cleanup - meshes

justinlaughlin committed:
Merge branch 'justinlaughlin/iga-elasticity-pa' of github.com:mfem/mfem into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Refactor SplineIntegrationRules to be more general

justinlaughlin committed:
Fix bug with Diffusion Kernel patch PA data

justinlaughlin committed:
Use Vector for weights

justinlaughlin committed:
add Array::Prod

justinlaughlin committed:
Use PatchBasisInfo in Diffusion Kernel + lots of refactoring/cleanup

justinlaughlin committed:
Improve speed of patch PA diffusion kernel

justinlaughlin committed:
Diffusion patch PA kernel optimization attempt

justinlaughlin committed:
Cleanup

justinlaughlin committed:
More cleanup

justinlaughlin committed:
Organize

justinlaughlin committed:
Refactor + cleanup

justinlaughlin committed:
Cleanup

justinlaughlin committed:
Cleanup + fix patch matrix assembly bug

justinlaughlin committed:
Use generalized patch G/GT for diffusion

justinlaughlin committed:
Cleanup + Add unique knot / multiplicity calculation

justinlaughlin committed:
Cleanup nurbs_patch_ex2

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Cleanup and syntax fixes

justinlaughlin committed:
Fix include paths for make

justinlaughlin committed:
Fix compiler warning

justinlaughlin committed:
More cleanup / fixes for compiler warnings

justinlaughlin committed:
Update gitignore

justinlaughlin committed:
remove unused variables

justinlaughlin committed:
More fixes for tests

justinlaughlin committed:
Hopefully fixes shadow variable issue on macos + fix mistake in nurbs_patch_ex1

justinlaughlin committed:
More shadow variables fixes for macos...

justinlaughlin committed:
More fixes

justinlaughlin committed:
Fixes

justinlaughlin committed:
Fix nurbs_patch_ex2

justinlaughlin committed:
Fix nurbs_patch_ex1

justinlaughlin committed:
Some styling + missing template parameter

justinlaughlin committed:
Typing

justinlaughlin committed:
Typing

justinlaughlin committed:
Typing

justinlaughlin committed:
Minor

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   cout << "Solving linear system ... " << endl;
+   solver.SetMaxIter(1e5);
+   solver.SetPrintLevel(1);
+   // These tolerances end up getting squared
```
This comment about the use of tolerances in `CGSolver` seems misleading. Should it be kept? In `CGSolver::Mult`, tolerances are squared to compute `r0`, which is used as a threshold for dot products of vectors. Thus the tolerance is squared, for comparison to squared quantities.

justinlaughlin replied:
I just removed the comment since this `sqrt` pattern is also in other examples

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+                  "Integration rule type: 0 - full Gaussian, "
+                  "1 - reduced Gaussian");
+   args.AddOption(&preconditioner, "-pc", "--preconditioner",
+                  "Preconditioner: 0 - none, 1 - diagonal");
```
`diagonal` -> `Jacobi`?

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+
+   // 12. Solve the linear system A X = B.
+   cout << "Solving linear system ... " << endl;
+   solver.SetMaxIter(1e5);
```
`1e5` is very high. For the sample runs, `1e4` would be sufficient.

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   MFEM_VERIFY(mesh.GetNodes(), "NURBS mesh must have nodes");
+   if (mesh.bdr_attributes.Max() < 2)
+   {
+      cout << "\nInput mesh should have at least two boundary"
```
Comparing to `ex2`, should there also be a check for two material attributes? Shouldn't the example terminate in this case?

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   // 4. Refine the mesh to increase the resolution.
+   for (int l = 0; l < ref_levels; l++)
+   {
+      // mesh.UniformRefinement();
```
Remove?

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   for (int l = 0; l < ref_levels; l++)
+   {
+      // mesh.UniformRefinement();
+      mesh.NURBSUniformRefinement();
```
This is fine, but it is also possible to refine once by a factor instead of using multiple refinements. One refinement should be more efficient, and it would be informative to demonstrate this in the examples.

justinlaughlin replied:
Right now `ref`/`ref_levels` is the same as `nurbs_patch_ex1` - should we change both to eg `rf`/`ref_factor`?

justinlaughlin replied:
Running `nurbs_patch_ex1` with `NURBSUniformRefinement(rf=2)` gives me a memory error 🤔 ; I can run other `rf` (3, 4, ...) just fine.

justinlaughlin replied:
Added in 790ee37

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   FiniteElementSpace *fespace = new FiniteElementSpace(&mesh, mesh.NURBSext, fec,
+                                                        dim, Ordering::byVDIM);
+   cout << "Finite Element Collection: " << fec->Name() << endl;
+   const int Ndof = fespace->GetTrueVSize();
```
For style consistency, we normally would use `ndof` in MFEM.

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,317 @@
+   FiniteElementCollection * fec = mesh.GetNodes()->OwnFEC();
+   cout << "fec order = " << fec->GetOrder() << endl;
+
+   FiniteElementSpace *fespace = new FiniteElementSpace(&mesh, mesh.NURBSext, fec,
```
`fespace` does not get deleted. It could just be an object, `FiniteElementSpace fespace ...`

dylan-copeland commented on fem/integ/bilininteg_patch.hpp:
```diff
@@ -0,0 +1,181 @@
+{
+
+/**
+ * Compute gradient at quadrature points
```
Mention that this is for NURBS patch integration?

justinlaughlin replied:
Done

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -52,6 +52,15 @@ class KnotVector
    int NumOfElements;
 
+   /// Stores the values of unique knots.
+   mutable Vector uknot;
```
Can we avoid making these `mutable` by calling `ComputeUniqueKnots` from `GetElements`?

dylan-copeland replied:
Yes, I agree that `ComputeUniqueKnots` should be called when knots are initialized or changed and should be called at the end of `KnotVector::Flip`. If `Flip` changes the knots but not the number of elements, then we can just call `ComputeUniqueKnots` at the end of `Flip`. Also, `GetElements` can always call `ComputeUniqueKnots`. If we don't need the unique knots, it is a waste of work, but it seems negligible (1D complexity).

justinlaughlin replied:
Added in [2ffd697](https://github.com/mfem/mfem/pull/4727/commits/2ffd697801dcc4cc3c9ac60f769ef1434a29c566)

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -168,6 +177,18 @@ class KnotVector
+   void GetUniqueKnots(Vector &uknots) const;
+
+   /// Get the ith unique knot's multiplicity.
+   real_t GetKnotMult(int i) const;
```
Return value of type `int`?

justinlaughlin replied:
Changed, good catch thank you

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -52,6 +52,15 @@ class KnotVector
+   mutable Vector uknot;
+
+   /// Stores the unique knot multiplicities.
+   mutable Vector uknot_mult;
```
`Array<int>`?

dylan-copeland commented on fem/bilinearform_ext.cpp:
```diff
@@ -441,7 +476,10 @@ void PABilinearFormExtension::AssembleDiagonal(Vector &y) const
    const int iSz = integrators.Size();
-   if (elem_restrict && !DeviceCanUseCeed())
+
+   bool patchwise = ContainsPatchwiseIntegration(integrators);
```
`const bool`

justinlaughlin commented on general/array.hpp:
```diff
@@ -311,6 +311,9 @@ class Array
    T Sum() const;
 
+   /// Return the product of all the array entries using the '*' operator for class 'T'.
+   T Prod() const;
```
I like this better too - changed

justinlaughlin committed:
Address most reviewer comments

justinlaughlin commented on fem/intrules.hpp:
```diff
@@ -268,12 +268,52 @@ class IntegrationRule : public Array<IntegrationPoint>
+       "Efficient and robust quadratures for isogeometric analysis:
+        Reduced Gauss and Gauss-Greville Rules" - Zou 2022
+       Generally, this rule should only be used as an approximation */
+   static IntegrationRule* GetReducedGaussianRule(const KnotVector &kv);
```
I was thinking `GetReducedGaussGrevilleRule` but this function (eq 14) doesnt use the greville points. We could call it `GetReducedGaussZouRule`?

justinlaughlin replied:
Or `GetReducedGaussLegendreRule`. We have other reduced rules but this one is just a simple Gauss Legendre with reduced points based on knot multiplicity

justinlaughlin committed:
Address more comments

justinlaughlin committed:
Address more comments - rename SplineIntegrationRule types

justinlaughlin committed:
Add ComputeUniqueKnots to constructors

justinlaughlin committed:
Some changes to patch integration - make sure to use max(q1d,d1d) for accumulators

justinlaughlin committed:
Change ref_levels to refinement_factor in nurbs_patch_ex1 and nurbs_patch_ex2

justinlaughlin committed:
Add tests for nurbs_patch_ex2

justinlaughlin commented on miniapps/nurbs/CMakeLists.txt:
```diff
@@ -40,6 +40,10 @@ add_mfem_miniapp(nurbs_ex1
+add_mfem_miniapp(nurbs_patch_ex2
+  MAIN nurbs_patch_ex2.cpp
+  LIBRARIES mfem)
+
```
Added in 790ee37 and 0f78cb55

justinlaughlin committed:
Fix typo

justinlaughlin committed:
Fix edge case of ComputeUniqueKnots. Add small comment to FixedOrderGaussLegendre

justinlaughlin committed:
Preconditioner leak

justinlaughlin committed:
Preconditioner leak

justinlaughlin committed:
Recompute unique knots on Flip()

justinlaughlin committed:
fix typo + cleanup

justinlaughlin committed:
Update nurbs_patch_ex2 test cases

justinlaughlin commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,314 @@
+   // 2. Read the mesh from the given mesh file.
+   Mesh mesh(mesh_file, 1, 1);
+   const int dim = mesh.Dimension();
+   const bool isNURBS = mesh.NURBSext;
```
This was a nice change in #4326 (see 97a9330) but it looks like it is gone now. I'm okay adding in back in (and using it throughout the code); what do you think?

justinlaughlin committed:
Implement more reviewer suggestions. add+use mesh.IsNURBS()

justinlaughlin commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,314 @@
+   const bool isNURBS = mesh.NURBSext;
+
+   // Verify mesh is valid for this problem
+   MFEM_VERIFY(isNURBS, "Example is for NURBS meshes");
```
added in 0b8f57d

justinlaughlin commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,314 @@
+   }
+
+   // 3. Optionally, increase the NURBS degree.
+   if (nurbs_degree_increase>0)
```
added in 0b8f57d

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

dylan-copeland commented on fem/bilininteg.hpp:
```diff
@@ -63,6 +64,8 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
 
    virtual void AssemblePABoundaryFaces(const FiniteElementSpace &fes);
 
+   virtual void AssembleDiagonalNURBSPA(Vector &diag);
```
Document that the diagonal is added to `diag` (copy comment from `AssembleDiagonalPA`).

dylan-copeland commented on mesh/nurbs.cpp:
```diff
@@ -20,6 +20,7 @@
 
 #include <fstream>
 #include <algorithm>
+#include <utility>
```
Why is this new include needed?

dylan-copeland commented on fem/integrator.hpp:
```diff
@@ -113,6 +113,43 @@ class Integrator
 };
+
+/** @brief This struct contains information about basis functions
+    used in patchwise integration. */
```
`patch-wise integration on NURBS meshes` in case someone is unaware that this is for NURBS.

justinlaughlin committed:
Address more reviewer comments

justinlaughlin committed:
Merge branch 'justinlaughlin/iga-elasticity-pa' of github.com:mfem/mfem into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Forgot endif

justinlaughlin committed:
Update fem/bilinearform_ext.cpp

Co-authored-by: Yohann <dudouit1@llnl.gov>

justinlaughlin commented on fem/bilinearform_ext.cpp:
```diff
@@ -38,6 +38,31 @@ const Operator *BilinearFormExtension::GetRestriction() const
+   }
+
+   MFEM_VERIFY(!(somePatchwise && !allPatchwise),
+               "All or none of the integrators should be patchwise");
```
Good idea

justinlaughlin committed:
Address Yohann's comments; clear up some logic with patchwise integration

justinlaughlin commented on fem/bilinearform_ext.cpp:
```diff
@@ -418,7 +443,17 @@ void PABilinearFormExtension::AssembleDiagonal(Vector &y) const
                                              Vector &d)
    {
-      integ.AssembleDiagonalPA(d);
+      if (integ.Patchwise())
```
[ab72ca4](https://github.com/mfem/mfem/pull/4727/commits/ab72ca43a2f1fc18a34379c25fc6c2dd8770824a)

justinlaughlin commented:
Hopefully [ab72ca4](https://github.com/mfem/mfem/pull/4727/commits/ab72ca43a2f1fc18a34379c25fc6c2dd8770824a) addresses your suggestions Yohann, thanks for the feedback! Once the CI is working again I'll re-run the tests...

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Patchwise logic; fix edge case (no integrators)

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Fix merge with nc nurbs changes

dylan-copeland commented on miniapps/nurbs/nurbs_patch_ex2.cpp:
```diff
@@ -0,0 +1,312 @@
+   FiniteElementCollection * fec = mesh.GetNodes()->OwnFEC();
+   cout << "fec order = " << fec->GetOrder() << endl;
+
+   FiniteElementSpace fespace = FiniteElementSpace(&mesh, fec,
```
Don't need `=`?

justinlaughlin replied:
Thank you, see 725fded9

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin committed:
Address reviewer comments

justinlaughlin committed:
Style

justinlaughlin committed:
Merge branch 'master' into justinlaughlin/iga-elasticity-pa

justinlaughlin commented:
> @justinlaughlin, can you please resolve the conflicts?

Resolved in 4dfae70e
